resdk/data_upload/samplesheet.py

```diff
@@ -0,0 +1,263 @@
+            self._read_xlrd()
+        elif self.extension() in ['.txt', '.tab', '.tsv']:
+            self._read_text_file()
+        else:
+            raise TypeError(
+                "Annotation spreadsheet extension `{}` not recognised. Options"
+                " are: `.xls`, `.xlsx`, '.xlsm', "
+                "`.txt`, `.tab`, `.tsv`.".format(self.extension())
+            )
+
+class Sample(object):
+    """Create a Sample like object."""
+
+    def __init__(self, entry):
+        self.entry = entry
+        self.validate()
+
+        # Extract the sample meta-data
+        self.name = entry['SAMPLE_NAME']
+        self.collection = entry['COLLECTION']
+        self.path = entry['FASTQ_R1']
+        self.path2 = entry['FASTQ_R2']
+        self.seq_type = entry['SEQ_TYPE']
+        self.reads_annotation = {
+            'experiment_type': self.seq_type,
+            'protocols': {
+                'extract_protocol': entry['EXTRACTION_PROTOCOL'],
+                'library_prep': entry['LIBRARY_CONSTRUCTION_PROTOCOL'],
+                'treatment_protocol': entry['TREATMENT_PROTOCOL'],
+                'growth_protocol': entry['GROWTH_PROTOCOL'],
+                'antibody_information': {
+                    'manufacturer' : entry['ANTIBODY']
+                }
+            },
+            'reads_info': {
+                'barcode': entry['BARCODE'],
+                'facility': entry['FACILITY']
+            }
+        }
+        self.molecule = entry['MOLECULE']
+        self.organism = entry['ORGANISM']
+        self.annotator = entry['ANNOTATOR']
+        self.source = entry['SOURCE']
+        self.sample_annotation = {
+            'sample': {
+                'cell_type': entry['CELL_TYPE'],
+                'strain': entry['STRAIN'],
+                'genotype': entry['GENOTYPE'],
+                'optional_char': []
+            }
+        }
+
+        # Include only if they are non-empty, to not override error-checking
+        for label, info in [
+                ('organism', self.organism),
+                ('molecule', self.molecule),
+                ('annotator', self.annotator),
+                ('source', self.source),
+            ]:
+            if info:
+                self.sample_annotation['sample'][label] = info
+
+        # Include optional columns
+        for option in sorted(OPTIONAL):
+            if entry[option]:
+                self.sample_annotation['sample']['optional_char'].append(
+                    '{0}:{1}'.format(option, entry[option]))
+
+    def validate(self):
+        """Validate the annotation spreadsheet file."""
+        # Check column headers
+        diff1 = COLUMNS - set(self.entry.keys())
+        diff2 = set(self.entry.keys()) - COLUMNS
+        err_msg = (
+            "Headers `{0}` {1}. You should use the"
+            " headers from Annotation_spreadsheet.xlsm found on "
+            "https://github.com/genialis/resolwe-bio-py/tree/master"
```

Comment:
This doesn't support versioning. I don't like the idea of downloading templates from github at all.

Reply:
Would it be better if this were a data object hosted in a public collection on one of the production servers?